src/net.cpp

```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
     elements.erase(std::remove_if(elements.end() - eraseSize, elements.end(), predicate), elements.end());
 }
 
-void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvictionCandidates)
+void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& candidates)
 {
+    // Networks to protect: localhost, onion, I2P. In case of equal counts, earlier array members
+    // have the first opportunity to recover unused protection slots from the previous iteration.
+    struct Net { Network id; size_t count; bool is_local; };
+    std::array<Net, 3> networks{{{/* localhost */ NET_MAX, 0, 1}, {NET_ONION, 0, 0}, {NET_I2P, 0, 0}}};
+
+    for (Net& n : networks) { // Count and store the number of eviction candidates per network.
+        n.count = std::count_if(candidates.cbegin(), candidates.cend(),
+                                [&n](const NodeEvictionCandidate& c) { return n.is_local ? c.m_is_local : c.m_network == n.id; });
+    }
+    // Sort `networks` by descending candidate count, to give networks having more candidates
+    // the first opportunity to recover unused protection slots from the previous iteration.
+    std::stable_sort(networks.begin(), networks.end(), [](Net a, Net b) { return a.count > b.count; });
```

Comment:
I think I would favor sorting network by ascending candidate count, otherwise I believe it would bias towards anonymity networks with abundant peers, decreasing our local diversity of anonymity network ?

I think this point has been already discussed in this [thread](https://github.com/bitcoin/bitcoin/pull/21261#discussion_r647319655). I agree it's minor.

Reply:
Done. As the unit tests show, one peer can make a difference when the numbers are low, e.g. when -maxconnections has been reduced or only one candidate will be protected.